README.md
# Preferences Inheritance

VS Code extension that allows sharing common preferences across profiles, workspaces, and projects through inheritance. It even allows fetching preferences from the web.

## Features

More precisely, this extension merges the given source of config files into a final config file that can be placed wherever you want.

### Example Scenario

**Config:**
```jsonc
{
    "targets": [
        {
            "sources": [
                "https://url.to/online/root/settings.json",
                {
                    "path": "path/to/local/base/settings.json",
                    "type": "jsonc"
                },
                "path/to/c++/project/settings.json"
            ],
            "output": "path/to/my/profile/workspace/.vscode/settings.json"
        }
    ],
    "notifyOnMerge": true,
    "mergeOnStartup": true,
    "watchSources": true
}
```

This config can be placed at the workspace, .vscode, or profile level. The extension will merge the sources into a final output config that gets applied. This way, I keep my setting files DRY and can reuse or combine them across multiple projects and environments however I want.


## Requirements

The extension won't do anything before specifying `preferencesInheritance.targets`.

## Extension Settings

This extension contributes the following settings:

* `preferencesInheritance.targets`:

    Example:

    `output: { path: "out.json", type: "json" }` // for output, type can be [json, text, yaml]

    or, just a string, will derive type from extension or default to json

    `output: "out.json"`

    `sources: [{ path: "test1.json", type: "jsonc" }, ...]` // for source, type can be [json, jsonc, text, yaml]
    or, just a string, content is merged (or concatenated in case of `type=text`) first to last.

    `sources: ["test1.json", { path: "test2.txt", type: "text"}, ...]`

* `preferencesInheritance.mergeOnStartup`: Merge preferences on startup.
* `preferencesInheritance.notifyOnMerge`: Notify when preferences are merged.
* `preferencesInheritance.notifyOnNewWorkspace`: Notify about merges being active (if list of merge sources is non-empty) when a new workspace is opened.
* `preferencesInheritance.watchSources`: Watch sources for changes and merge on change.
* `preferencesInheritance.fetchContentCacheTime`: Time in seconds to cache fetched web content. Default is 60 * 5 seconds.

## Release Notes


### 1.0.0

First release


CHANGELOG.md
# Change Log

All notable changes to the "preferences-inheritance" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

.vscode-test.mjs
import { defineConfig } from "@vscode/test-cli";

export default defineConfig({
  files: "out/test/**/*.test.js",
  version: "1.60.0",
  launchArgs: ["--disable-extensions"],
  mocha: {},
});


src/test/extension.test.ts
import * as assert from "assert";
import { after, afterEach, beforeEach } from "mocha";
import sinon from "sinon";
import { dir } from "tmp-promise";
import * as vscode from "vscode";

import pkg from "../../package.json";
import { MergeObject } from "../constants";
import { deactivate } from "../extension";
import { loadSettings, SettingsInheritance } from "../inheritance";

function setPreferences(prefs: Record<string, any>) {
  sinon.replace(
    vscode.workspace,
    "getConfiguration",
    () =>
      ({
        get: (key: string, def: any) => prefs[key] ?? def,
      } as any)
  );
}

function createThenable<T>(value: T) {
  return {
    then: async (cb: (value: T) => void) => {
      await cb(value);
    },
  };
}

suite("Extension Test Suite", () => {
  vscode.window.showInformationMessage("Start all tests.");
  after(() => {
    vscode.window.showInformationMessage("All tests done!");
  });

  const ext = vscode.extensions.getExtension(`${pkg.publisher}.${pkg.name}`);

  let context: vscode.ExtensionContext;
  let inheritance: SettingsInheritance;
  let tmpdir: string;
  let cleanup: () => Promise<void>;

  let prefs: Record<string, any>;

  let mergePrefs = [
    {
      sources: [
        {
          path: "test1.json",
          type: "json",
        },
        {
          path: "test2.json",
          type: "jsonc",
        },
        {
          path: "test3.txt",
          type: "text",
        },
      ],
      output: {
        path: "out.json",
        type: "json",
      },
    },
  ];

  beforeEach(async () => {
    const tmp = await dir({ unsafeCleanup: true });
    tmpdir = tmp.path;
    cleanup = tmp.cleanup;

    assert.ok(tmp);

    assert.ok(ext);

    const ctx = await ext!.activate();
    assert.ok(ctx);
    context = ctx.context;
    assert.ok(context);

    inheritance = new SettingsInheritance();
    assert.ok(inheritance);

    prefs = {
      targets: [
        {
          sources: [
            {
              path: "test1.json",
              type: "json",
            },
            "test2.json",
            {
              path: "test3.txt",
            },
          ],
          output: "out.json",
        },
      ],
      notifyOnMerge: true,
      notifyOnNewWorkspace: true,
      mergeOnStartup: true,
      watchSources: true,
      fetchContentCacheTime: 1000,
    };

    setPreferences(prefs);
  });

  afterEach(async () => {
    if (inheritance) {
      await inheritance.deactivate();
    }

    if (tmpdir) {
      await cleanup();
    }
    sinon.restore();
  });

  test("load merge settings", async () => {
    const s = loadSettings();

    assert.deepStrictEqual(s, { ...prefs, targets: mergePrefs });
  });

  test("merge settings json", async () => {
    const obj: MergeObject = {
      sources: [
        {
          path: "first.json",
          type: "json",
        },
        {
          path: "second.json",
          type: "jsonc",
        },
      ],
      output: {
        path: "out.json",
        type: "json",
      },
    };

    sinon.replace(vscode.workspace, "openTextDocument", ((path: string) => {
      let r: any;
      if (path === "first.json") {
        r = {
          getText: () => '{"first": 1}',
        } as any;
      } else if (path === "second.json") {
        r = {
          getText: () => `{"second": 2} // comment`,
        } as any;
      } else if (path === "out.json") {
        r = {
          getText: () => "",
          save: async () => true,
        } as any;
      } else {
        throw new Error("Unknown path");
      }

      r = {
        ...r,
        uri: vscode.Uri.file(path),
        positionAt: () => new vscode.Position(0, 0),
      };

      return createThenable(r);
    }) as any);

    const f1 = sinon.fake.resolves(true);
    const f2 = sinon.fake();
    sinon.replace(vscode.workspace, "applyEdit", f1);
    sinon.replace(vscode.WorkspaceEdit.prototype, "replace", f2);
    await inheritance.merge(obj);

    assert.ok(f1.calledOnce);
    assert.ok(f2.calledOnce);

    assert.equal(f2.firstCall.args[0].path, "/out.json");
    assert.equal(f2.firstCall.args[2], '{\n  "first": 1,\n  "second": 2}');
  });

  test("merge settings text", async () => {
    const obj: MergeObject = {
      sources: [
        {
          path: "first.json",
          type: "json",
        },
        {
          path: "second.json",
          type: "plaintext",
        },
      ],
      output: {
        path: "out.txt",
        type: "text",
      },
    };

    sinon.replace(vscode.workspace, "openTextDocument", ((path: string) => {
      let r: any;
      if (path === "first.json") {
        r = {
          getText: () => '{"first": 1}',
        } as any;
      } else if (path === "second.json") {
        r = {
          getText: () => `{"second": 2} // comment`,
        } as any;
      } else if (path === "out.txt") {
        r = {
          getText: () => "",
          save: async () => true,
        } as any;
      } else {
        throw new Error("Unknown path");
      }

      r = {
        ...r,
        uri: vscode.Uri.file(path),
        positionAt: () => new vscode.Position(0, 0),
      };

      return createThenable(r);
    }) as any);

    const f1 = sinon.fake.resolves(true);
    const f2 = sinon.fake();
    sinon.replace(vscode.workspace, "applyEdit", f1);
    sinon.replace(vscode.WorkspaceEdit.prototype, "replace", f2);
    await inheritance.merge(obj);

    assert.ok(f1.calledOnce);
    assert.ok(f2.calledOnce);

    assert.equal(f2.firstCall.args[0].path, "/out.txt");
    assert.equal(
      f2.firstCall.args[2],
      '"\\"\\"{\\"first\\":1}\\"{\\\\\\"second\\\\\\": 2} // comment\\""'
    );
  });

  test("merge settings web url", async () => {
    const obj: MergeObject = {
      sources: [
        {
          path: "https://test.com/first.json",
          type: "json",
        },
        {
          path: "second.json",
          type: "json",
        },
      ],
      output: {
        path: "out.json",
        type: "json",
      },
    };

    sinon.replace(vscode.workspace, "openTextDocument", ((path: string) => {
      let r: any;
      if (path === "second.json") {
        r = {
          getText: () => `{"second": 2} // comment`,
        } as any;
      } else if (path === "out.json") {
        r = {
          getText: () => "",
          save: async () => true,
        } as any;
      } else {
        throw new Error("Unknown path");
      }

      r = {
        ...r,
        uri: vscode.Uri.file(path),
        positionAt: () => new vscode.Position(0, 0),
      };

      return createThenable(r);
    }) as any);

    const f1 = sinon.fake.resolves(true);
    const f2 = sinon.fake();
    sinon.replace(vscode.workspace, "applyEdit", f1);
    sinon.replace(vscode.WorkspaceEdit.prototype, "replace", f2);

    sinon.replace(global, "fetch", async () => {
      return {
        text: async () => '{"web": 1}',
        ok: true,
      } as any;
    });

    await inheritance.merge(obj);

    assert.ok(f1.calledOnce);
    assert.ok(f2.calledOnce);

    assert.equal(f2.firstCall.args[0].path, "/out.json");
    assert.equal(f2.firstCall.args[2], '{\n  "web": 1,\n  "second": 2}');
  });
});


src/utils.ts
import JSON5 from 'json5';
import * as vscode from 'vscode';
import * as yaml from 'yaml';

import {
  constant,
  DocumentOutputType,
  DocumentType,
  JSONMappable,
  JSONMapType,
  JSONType,
} from './constants';

export function basename(path: string) {
  const parts = path.split(/[\/\\]/);
  return [parts[parts.length - 1], parts.slice(0, parts.length - 1).join("/")];
}

export function debounce<T extends (...args: any[]) => any>(
  fn: T,
  delay: number
): T {
  let timeoutID: NodeJS.Timeout | undefined;

  return ((...args: any[]) => {
    if (timeoutID) {
      clearTimeout(timeoutID);
    }

    timeoutID = setTimeout(() => {
      fn(...args);
    }, delay);
  }) as any;
}

export function showDocument(
  path: string,
  content?: undefined,
  type?: DocumentType
): Promise<boolean>;
export function showDocument(
  path: string | undefined,
  content: string,
  type: DocumentType
): Promise<boolean>;
export function showDocument(
  path: string | undefined,
  content: string | undefined,
  type: DocumentType | undefined
): Promise<boolean> {
  return new Promise<boolean>((resolve) => {
    const reject = () => resolve(false);

    const lang = type
      ? (
          {
            text: "plaintext",
          } as any
        )[type] ?? type
      : undefined;

    if (path && content) {
      vscode.workspace
        .openTextDocument({
          content,
          language: lang,
        })
        .then((doc) => {
          if (lang) {
            vscode.languages.setTextDocumentLanguage(doc, lang).then((doc) => {
              vscode.window.showTextDocument(doc);
            }, reject);
          } else {
            vscode.window.showTextDocument(doc);
          }
          resolve(true);
        }, reject);
    } else if (path) {
      vscode.workspace.openTextDocument(path).then((doc) => {
        vscode.window.showTextDocument(doc);
        resolve(true);
      }, reject);
    } else {
      resolve(false);
    }
  });
}

export function parseContent(text: string, type: DocumentType): JSONType {
  switch (type) {
    case "json":
    case "jsonc":
      return JSON5.parse(text) as JSONType;
    case "yaml":
      return yaml.parse(text) as JSONType;
    case "plaintext":
    case "text":
      return text as JSONType;
    default:
      throw new Error(`Unknown type "${type}"`);
  }
}

export function isObject(obj: any): obj is Record<any, any> {
  return obj instanceof Object && !(obj instanceof Array);
}

// like JSON.stringify but keeps the insertion order of the keys in case of a Map
export function stringifyMap(
  value: Record<string, any> | JSONMapType,
  stringify: (value: any) => string,
  opts?: {
    indent?: number;
  },
  _depth = 1
) {
  if (value instanceof Map) {
    const indent = (opts?.indent ?? 0) * _depth;
    const spaces = " ".repeat(indent);
    let out = "{";
    let i = 0;
    for (const [k, v] of value.entries()) {
      const innerValue =
        v instanceof Map
          ? stringifyMap(v, stringify, opts, _depth + 1)
          : stringify(v);

      out += `\n${spaces}"${k}": ${innerValue}`;

      i++;
      if (i < value.size) {
        out += ",";
      }
    }
    out += "}";
    return out;
  } else {
    return stringify(value);
  }
}

export function stringifyContent(
  value: string | JSONType | JSONMappable,
  type: DocumentOutputType
) {
  const indent = constant.JSONIndent;

  switch (type) {
    case "plaintext":
    case "text":
      return JSON.stringify(value);
    case "json":
    case "yaml": {
      if (value instanceof Map) {
        switch (type) {
          case "json":
            return stringifyMap(value, (v) => JSON.stringify(v, null, indent), {
              indent,
            });
          case "yaml":
            return stringifyMap(value, (v) => yaml.stringify(v, { indent }), {
              indent,
            });
        }
      } else {
        switch (type) {
          case "json":
            return JSON.stringify(value, null, indent);
          case "yaml":
            return yaml.stringify(value, { indent });
        }
      }
    }
    default:
      throw new Error(`Unknown type "${type}"`);
  }
}

export function camelCase(str: string) {
  return str.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
}

export function isWebUrl(url: string | URL) {
  try {
    const u = typeof url === "string" ? new URL(url) : url;
    if (u.protocol === "http:" || u.protocol === "https:") {
      return true;
    }
  } catch (e) {}
  return false;
}

export function downloadDocument(url: URL, clean = true) {
  if (!isWebUrl(url.toString())) {
    throw new Error(`Invalid URL "${url}"`);
  }

  return fetch(url).then((res) => {
    if (!res.ok) {
      throw new Error(
        `Failed to download document from "${url}":\n${res.status} - ${res.statusText}`
      );
    }

    return res.text().then((text) => {
      if (clean) {
        return text.trim();
      }
      return text;
    });
  });
}


src/constants.ts
import * as vscode from "vscode";

import type { SettingsInheritance } from "./inheritance";

export type ValueOf<T> = T[keyof T];
export type JSONType =
  | string
  | number
  | boolean
  | null
  | JSONType[]
  | { [key: string]: JSONType };

export type JSONMapType = Map<string, JSONType | JSONMapType>;

export type JSONMappable = Record<string, JSONType | JSONMapType> | JSONMapType;

export interface Settings {
  targets: MergeObject[];
  notifyOnMerge: boolean;
  notifyOnNewWorkspace: boolean;
  mergeOnStartup: boolean;
  watchSources: boolean;
  fetchContentCacheTime: number;
}

export const documentType = [
  "jsonc",
  "text",
  "plaintext",
  "json",
  "yaml",
] as const;
export type DocumentType = (typeof documentType)[number];
export const documentOutputType = [
  "text",
  "plaintext",
  "json",
  "yaml",
] as const;
export type DocumentOutputType = (typeof documentOutputType)[number];

export interface MergeSource {
  path: string;
  type: DocumentType;
}

export interface MergeOutput {
  path: string;
  type: DocumentOutputType;
}

export interface MergeObject {
  sources: MergeSource[];
  output: MergeOutput;
}

export const EXT_NAME = "preferences-inheritance";

export const NoWorkspace = Symbol();
export type WorkspaceRootPath = string | typeof NoWorkspace;

export class constant {
  static inheritance: SettingsInheritance;
  static context: vscode.ExtensionContext;
  static settings: Settings = {} as Settings;

  static contentCache: Record<string, string> = {};
  static lastContentFetchTime = 0;
  static JSONIndent = 2;

  static readonly workspaceState: Record<
    WorkspaceRootPath,
    {
      notified: boolean;
    }
  > = {} as any;
}


src/extension.ts
import * as vscode from "vscode";

import { constant, EXT_NAME } from "./constants";
import { loadSettings, SettingsInheritance } from "./inheritance";

export async function activate(context: vscode.ExtensionContext) {
  console.log("Loading extension " + EXT_NAME);
  constant.context = context;

  loadSettings();

  if (constant.inheritance) {
    await constant.inheritance.deactivate();
  }

  constant.inheritance = new SettingsInheritance();
  await constant.inheritance.activate(context);

  return {
    context,
    activate,
    deactivate,
    inheritance: constant.inheritance,
  };
}

// This method is called when your extension is deactivated
export function deactivate() {
  if (constant.inheritance) {
    constant.inheritance.deactivate();
    // @ts-expect-error
    constant.inheritance = undefined;
  }
}


src/inheritance.ts
import * as vscode from "vscode";

import {
  constant,
  documentOutputType,
  documentType,
  EXT_NAME,
  JSONMappable,
  JSONType,
  MergeObject,
  MergeOutput,
  MergeSource,
  NoWorkspace,
  ValueOf,
} from "./constants";
import { deactivate } from "./extension";
import {
  basename,
  camelCase,
  debounce,
  downloadDocument,
  isObject,
  isWebUrl,
  parseContent,
  showDocument,
  stringifyContent,
  stringifyMap,
} from "./utils";

function showError(msg: string, e?: any, warning?: boolean) {
  const cb = warning
    ? vscode.window.showWarningMessage
    : vscode.window.showErrorMessage;
  if (e) {
    cb(`${msg}`, {});
  } else {
    cb(`${msg}: ${e?.message ?? e}`);
  }
  console.error(`Settings Inheritance: ${msg}`);
  if (e) {
    console.error(e);
  }
}

export function loadSettings() {
  const cfg = vscode.workspace.getConfiguration(camelCase(EXT_NAME));
  const defaultOutputType = "json" as const;
  const defaultSourceType = "jsonc" as const;

  function getType<T>(
    t: any,
    path: string,
    def: T,
    types: typeof documentType | typeof documentOutputType
  ): T {
    if (t) {
      return t as T;
    }

    if (path) {
      const ext = path.split(".").pop();
      if (types.includes(ext as any)) {
        return ext as T;
      }
      if (ext === "yml") {
        return "yaml" as T;
      }
      if (ext === "txt") {
        return "text" as T;
      }
    }

    return def;
  }

  function validateMergePrefs(value: any): MergeObject[] {
    let r: MergeObject[] = [];

    if (!Array.isArray(value)) {
      showError("Invalid merge settings, expected array", value);
      return r;
    }

    for (const v of value) {
      if (!isObject(v)) {
        showError("Invalid merge settings, expected object", v);
        continue;
      }

      const sources: MergeSource[] = [];
      for (const s of v?.sources ?? []) {
        if (!s) {
          continue;
        }
        if (typeof s === "string") {
          sources.push({
            path: s,
            type: defaultSourceType,
          });
        } else if (s?.path) {
          sources.push({
            path: s.path,
            type: getType(s.type, s.path, defaultSourceType, documentType),
          });
        }
      }

      const oPath = typeof v?.output === "string" ? v?.output : v?.output?.path;

      if (oPath && sources.length) {
        r.push({
          sources,
          output: {
            path: oPath,
            type: getType(
              v?.output?.type,
              oPath,
              defaultOutputType,
              documentOutputType
            ),
          },
        });
      }
    }

    return r;
  }

  constant.settings = {
    targets: validateMergePrefs(cfg.get<MergeObject[]>("targets", [])),
    notifyOnMerge: cfg.get<boolean>("notifyOnMerge", false),
    notifyOnNewWorkspace: cfg.get<boolean>("notifyOnNewWorkspace", true),
    mergeOnStartup: cfg.get<boolean>("mergeOnStartup", true),
    fetchContentCacheTime: cfg.get<number>(
      "fetchContentCacheTime",
      60 * 5 // 5 minutes
    ),
    watchSources: cfg.get<boolean>("watchSources", true),
  };

  return constant.settings;
}

export function getWorkspaceState<
  K extends keyof ValueOf<typeof constant.workspaceState>,
  T extends ValueOf<typeof constant.workspaceState>[K]
>(key: K, defaultValue?: T) {
  const rootPath =
    vscode.workspace.workspaceFolders?.[0].uri.fsPath ?? NoWorkspace;
  if (rootPath) {
    if (!constant.workspaceState[rootPath]) {
      constant.workspaceState[rootPath] = {
        notified: false,
      };
    }
    return constant.workspaceState[rootPath][key] ?? defaultValue;
  }
  return defaultValue;
}

export function setWorkspaceState<
  K extends keyof ValueOf<typeof constant.workspaceState>,
  T extends ValueOf<typeof constant.workspaceState>[K]
>(key: K, value: T) {
  const rootPath =
    vscode.workspace.workspaceFolders?.[0].uri.fsPath ?? NoWorkspace;
  if (rootPath) {
    getWorkspaceState(key);
    constant.workspaceState[rootPath][key] = value;
  }
  return value;
}

export class SettingsInheritance {
  _disposables: vscode.Disposable[];
  _watchDisposables: vscode.Disposable[];

  constructor() {
    this._disposables = [];
    this._watchDisposables = [];
  }

  private async saveOutput(path: string, content: string) {
    return new Promise<boolean>((resolve, reject) => {
      vscode.workspace.openTextDocument(path).then(async (doc) => {
        let changed = false;
        if (doc.getText() !== content) {
          changed = true;
          const edit = new vscode.WorkspaceEdit();
          edit.replace(
            doc.uri,
            new vscode.Range(
              doc.positionAt(0),
              doc.positionAt(doc.getText().length)
            ),
            content
          );
          if (!(await vscode.workspace.applyEdit(edit))) {
            return reject("Failed to apply edit");
          }
          if (!(await doc.save())) {
            return reject("Failed to save output document");
          }
        }
        return resolve(changed);
      }, reject);
    });
  }

  private _mergeObject(
    obj: JSONMappable,
    source: JSONMappable,
    opts?: {
      mergeObjectDepth?: number;
    }
  ) {
    function setValue(k: string, v: any) {
      if (obj instanceof Map) {
        obj.set(k, v);
      } else {
        obj[k] = v;
      }
    }

    for (const [key, value] of source instanceof Map
      ? source.entries()
      : Object.entries(source)) {
      if (opts?.mergeObjectDepth && (value instanceof Map || isObject(value))) {
        for (const [k, v] of value instanceof Map
          ? value.entries()
          : Object.entries(value)) {
          const sourceValue = obj instanceof Map ? obj.get(key) : obj[key];

          const vIsObj = v instanceof Map || isObject(v);
          const sourceValueIsObj =
            sourceValue instanceof Map || isObject(sourceValue);

          if (vIsObj) {
            setValue(
              k,
              this._mergeObject(
                sourceValueIsObj ? (sourceValue as JSONMappable) : new Map(),
                v as JSONMappable,
                {
                  mergeObjectDepth: opts.mergeObjectDepth - 1,
                }
              )
            );
          } else {
            setValue(k, v);
          }
        }
      } else {
        setValue(key, value);
      }
    }
    return obj;
  }

  mergeObject(obj: JSONMappable, ...sources: JSONMappable[]) {
    for (const source of sources) {
      obj = this._mergeObject(obj, source, {
        mergeObjectDepth: 1, // TODO: configurable?
      });
    }
    return obj;
  }

  mergeContent(content: Map<MergeSource, string>, output: MergeOutput) {
    let prefsType = "map" as "map" | "text";

    if (["plaintext", "text"].includes(output.type)) {
      prefsType = "text";
    }

    let prefs: JSONType | JSONMappable = prefsType === "text" ? "" : new Map();

    // Map remembers the original insertion order of the keys so this is ok
    for (const [source, txt] of content.entries()) {
      const v = parseContent(txt, source.type);

      switch (source.type) {
        case "plaintext":
        case "text": {
          if (prefsType !== "text") {
            prefs =
              prefs instanceof Map && !prefs.size ? "" : JSON.stringify(prefs);
            const e =
              "Source with type='text' found when output has type='json'. Concatenating merge strategy will be used.";
            console.warn("Preferences Inheritance: " + e);
            showError(e, undefined, true);
          }
          prefsType = "text";
          prefs += JSON.stringify(v);
          break;
        }
        default: {
          if (prefsType === "text" || v === null) {
            prefs =
              prefs instanceof Map || isObject(prefs)
                ? stringifyMap(
                    prefs,
                    (v) => JSON.stringify(prefs, null, constant.JSONIndent),
                    { indent: constant.JSONIndent }
                  )
                : JSON.stringify(prefs);
            prefs += JSON.stringify(v);
          } else {
            if (v instanceof Map || isObject(v)) {
              prefs = this.mergeObject(
                prefs as JSONMappable,
                v as JSONMappable
              );
            } else {
              throw new Error(`Expected object. Cannot merge ${typeof prefs}`);
            }
          }
        }
      }
    }

    // double newlines
    "hello \n".replace(/\n/g, "\n\n");

    return stringifyContent(prefs, output.type);
  }

  async merge(obj: MergeObject) {
    let changed = false;
    const content: Map<MergeSource, string> = new Map();
    let outputErr: string = "";

    try {
      for (const source of obj.sources) {
        if (isWebUrl(source.path)) {
          const fetchTime = constant.settings.fetchContentCacheTime * 1000;
          if (
            !constant.contentCache[source.path] ||
            Date.now() - constant.lastContentFetchTime > fetchTime
          ) {
            const empty = Symbol();
            try {
              constant.contentCache[source.path] = await downloadDocument(
                new URL(source.path)
              );

              if (!constant.contentCache[source.path]) {
                throw empty;
              }
              constant.lastContentFetchTime = Date.now();
            } catch (e) {
              if (e === empty) {
                console.warn(
                  "Preferences Inheritance: Empty content from URL",
                  source.path
                );
                constant.lastContentFetchTime += 1000;
              } else {
                content.set(
                  source,
                  `<${(e as { message?: string })?.message ?? e}>`
                );
                constant.lastContentFetchTime += fetchTime * 0.25;
                throw e;
              }
            }
          }

          content.set(source, constant.contentCache[source.path]);
        } else {
          let txt: string | undefined;

          await vscode.workspace.openTextDocument(source.path).then(
            (doc) => {
              txt = doc.getText().trim();
            },
            (err) => {
              console.error(err);
              throw err;
              return undefined;
            }
          );

          if (txt) {
            content.set(source, txt);
          } else {
            console.warn(
              `Preferences Inheritance: Empty content or error from file`,
              source.path
            );
          }
        }
      }

      try {
        const output = this.mergeContent(content, obj.output);
        changed = await this.saveOutput(obj.output.path, output);
      } catch (e) {
        outputErr = `<${(e as { message?: string })?.message ?? e}>`;
        throw e;
      }
    } catch (e) {
      vscode.window
        .showErrorMessage(
          `Failed to merge settings. Open Developer Tools for more details."\n${e}`,
          "Show sources & output",
          "Show only output"
        )
        .then((value) => {
          switch (value) {
            case "Show sources & output": {
              for (const source of obj.sources) {
                const v = content.get(source);
                if (v) {
                  showDocument(source.path, v, source.type);
                }
              }

              showDocument(obj.output.path, outputErr, obj.output.type);
              break;
            }
            case "Show only output":
              showDocument(obj.output.path, outputErr, obj.output.type);
              break;
          }
        });
      console.error("Failed to merge settings");
      console.error(e);
    }

    let notified = false;
    if (constant.settings.notifyOnNewWorkspace) {
      if (!getWorkspaceState("notified", false)) {
        vscode.window.showInformationMessage(
          `Preferences Inheritance is now active for this workspace.`
        );
        notified = true;
        setWorkspaceState("notified", true);
      }
    }

    if (!notified && changed) {
      const txt =
        "Settings were merged. Please reload VS Code to apply changes.";

      if (constant.settings.notifyOnMerge) {
        vscode.window.showInformationMessage(txt, "Reload").then((value) => {
          if (value === "Reload") {
            vscode.commands.executeCommand("workbench.action.reloadWindow");
          }
        });
      } else {
        vscode.window.setStatusBarMessage(txt, 5000);
      }
    }
  }

  watch(obj: MergeObject) {
    const sources = [...obj.sources];

    const cb = debounce(() => {
      this.merge(obj);
    }, 1000);

    for (const source of sources) {
      if (isWebUrl(source.path)) {
        cb();
      } else {
        const [name, head] = basename(source.path);

        const p = new vscode.RelativePattern(vscode.Uri.file(head), name);

        const watcher = vscode.workspace.createFileSystemWatcher(
          p,
          false,
          false,
          false
        );

        this._watchDisposables.push(watcher.onDidChange(cb));
        this._watchDisposables.push(watcher.onDidCreate(cb));
        this._watchDisposables.push(watcher);
      }
    }
  }

  async start() {
    loadSettings();
    this._watchDisposables.forEach((d) => d.dispose());
    for (const obj of constant.settings.targets) {
      await this.merge(obj);
      if (constant.settings.watchSources) {
        this.watch(obj);
      }
    }
  }

  async activate(context: vscode.ExtensionContext) {
    this.dispose();

    context.subscriptions.push(
      vscode.commands.registerCommand(
        `${camelCase(EXT_NAME)}.mergePreferences`,
        () => {
          this.start();
        }
      )
    );

    if (constant.settings.mergeOnStartup) {
      await this.start();
    }
  }

  deactivate() {
    this.dispose();
  }

  dispose() {
    this._disposables.forEach((d) => d.dispose());
    this._watchDisposables.forEach((d) => d.dispose());
  }
}
